cloud-controller-manager/do/firewall_controller_diff.go

```diff
@@ -0,0 +1,127 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package do
+
+import (
+	"strings"
+
+	"github.com/digitalocean/godo"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+// comparableFirewall is a struct holding just the fields we care about during
+// comparison.
+type comparableFirewall struct {
+	Name          string
+	InboundRules  []godo.InboundRule
+	OutboundRules []godo.OutboundRule
+	Tags          []string
+}
+
+func firewallRequestEqual(fw *godo.Firewall, fr *godo.FirewallRequest) (bool, string) {
+	cf1 := compFirewallFromFirewall(fw)
+	cf2 := compFirewallFromFirewallRequest(fr)
+
+	return compFirewallsEqual(cf1, cf2)
+}
+
+func firewallsEqual(fw1, fw2 *godo.Firewall) (bool, string) {
+	cf1 := compFirewallFromFirewall(fw1)
+	cf2 := compFirewallFromFirewall(fw2)
+
+	return compFirewallsEqual(cf1, cf2)
+}
+
+func compFirewallsEqual(cf1, cf2 *comparableFirewall) (bool, string) {
+	// Equality when both variables are empty.
+	if cf1 == nil && cf2 == nil {
+		return true, ""
+	}
+
+	// Non-equality when exactly one of two variables is empty.
+	if cf1 == nil || cf2 == nil {
+		return false, ""
+	}
+
+	// Define custom sorters to guard against non-deterministic rule sort orders.
+	sorterInboundRules := cmpopts.SortSlices(func(r1, r2 godo.InboundRule) bool {
+		return printInboundRule(r1) < printInboundRule(r2)
+	})
+	sorterOutboundRules := cmpopts.SortSlices(func(r1, r2 godo.OutboundRule) bool {
+		return printOutboundRule(r1) < printOutboundRule(r2)
+	})
+
+	// A PortRange indicates "all ports" when the value is set to "all" on requests,
+	// and the corresponding response returns "0" then. Map them all to a single
+	// value for comparing.
+	portRangeMapper := cmp.FilterPath(func(p cmp.Path) bool {
+		switch p.String() {
+		case "InboundRules.PortRange", "OutboundRules.PortRange":
+			return true
+		}
+		return false
+	}, cmp.Comparer(func(pr1, pr2 string) bool {
+		sanitizePortRange := func(pr string) string {
+			if pr == "" || pr == "0" || pr == "all" {
+				pr = "0"
+			}
+			return pr
+		}
+
+		return sanitizePortRange(pr1) == sanitizePortRange(pr2)
+	}))
+
+	// Ignore all fields on {In,Out}boundRules.{Sources,Destinations} other than
+	// "Addresses".
+	ruleSourceDestFilter := cmp.FilterPath(func(p cmp.Path) bool {
+		if strings.HasPrefix(p.String(), "InboundRules.Sources") || strings.HasPrefix(p.String(), "OutboundRules.Destinations") {
+			return p.Last().String() != ".Addresses"
+		}
+
+		return false
+	}, cmp.Ignore())
+
+	diff := cmp.Diff(cf1, cf2, sorterInboundRules, sorterOutboundRules, portRangeMapper, ruleSourceDestFilter, cmpopts.EquateEmpty())
+	return diff == "", diff
+}
+
+func compFirewallFromFirewall(fw *godo.Firewall) *comparableFirewall {
+	if fw == nil {
+		return nil
+	}
+
+	return &comparableFirewall{
+		Name:          fw.Name,
+		InboundRules:  fw.InboundRules,
+		OutboundRules: fw.OutboundRules,
+		Tags:          fw.Tags,
```

Comment:
Is it possible for `fm.Tags` to return a nil slice and if so, could this be problematic if we don't check?

Reply:
With regards to accessing the value, it should not cause issues since we always read the value in a way that's unproblematic (e.g., use the `%s` placeholder with `fmt.Printf`).

With regards to the result of our comparison logic, it should also not cause issues since we pass `cmpopts.EquateEmpty()` to `cmp.Diff`, making sure that nil slices are considered equal to empty ones.